plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcModule.java

```diff
@@ -83,29 +84,43 @@ public void setup(Binder binder)
         bindSessionPropertiesProvider(binder, JdbcWriteSessionProperties.class);
         bindSessionPropertiesProvider(binder, JdbcDynamicFilteringSessionProperties.class);
 
-        binder.bind(DynamicFilteringStats.class).in(Scopes.SINGLETON);
+        binder.bind(DynamicFilteringStats.class).in(SINGLETON);
         newExporter(binder).export(DynamicFilteringStats.class)
                 .as(generator -> generator.generatedNameOf(DynamicFilteringStats.class, catalogName));
 
-        binder.bind(CachingJdbcClient.class).in(Scopes.SINGLETON);
-        binder.bind(JdbcClient.class).to(Key.get(CachingJdbcClient.class)).in(Scopes.SINGLETON);
+        binder.bind(CachingJdbcClient.class).in(SINGLETON);
+        binder.bind(JdbcClient.class).to(Key.get(CachingJdbcClient.class)).in(SINGLETON);
 
-        newSetBinder(binder, Procedure.class).addBinding().toProvider(FlushJdbcMetadataCacheProcedure.class).in(Scopes.SINGLETON);
+        newSetBinder(binder, Procedure.class).addBinding().toProvider(FlushJdbcMetadataCacheProcedure.class).in(SINGLETON);
 
         newSetBinder(binder, ConnectorTableFunction.class);
 
         binder.bind(ConnectionFactory.class)
                 .annotatedWith(ForLazyConnectionFactory.class)
                 .to(Key.get(ConnectionFactory.class, StatsCollecting.class))
-                .in(Scopes.SINGLETON);
-        binder.bind(ConnectionFactory.class).to(LazyConnectionFactory.class).in(Scopes.SINGLETON);
+                .in(SINGLETON);
+        install(ConditionalModule.conditionalModule(
+                BaseJdbcConfig.class,
+                BaseJdbcConfig::isReuseConnection,
+                innerBinder -> {
+                    innerBinder.bind(ConnectionFactory.class)
+                            .annotatedWith(ForReusingConnectionFactory.class)
+                            .to(Key.get(LazyConnectionFactory.class))
+                            .in(SINGLETON);
+                    innerBinder.bind(ReusingConnectionFactory.class).in(SINGLETON);
+                    innerBinder.bind(ConnectionFactory.class).to(Key.get(ReusingConnectionFactory.class)).in(SINGLETON);
```

Comment:
I think this should be enough:

```suggestion
                    innerBinder.bind(ConnectionFactory.class).to(ReusingConnectionFactory.class).in(SINGLETON);
```


but i cannot verify that since this code isn't exercised  by tests

Reply:
It is tested by `TestJdbcConnectionsCreation`. They `Key` is a key here, otherwise you would get two different instances.